### PR TITLE
[LLM] Skill for re-structuring and refactoring components

### PR DESCRIPTION
**Pull request by KenAJoh**
### Description

The skill, given a component-dir will try to refactor and re-structure it to a unified setup, with no breaking changes externally.
Tested it on Chips, and worked fine. 

Only "breaking" change from re-structuring is that props-sync for website no longer updates sine file-id is based on filepath. So will need, in this case to update the props-table for Chips docs after merge.

### Component Checklist 📝

- [ ] JSDoc
- [ ] Examples
- [ ] Documentation / Decision Records
- [ ] Storybook
- [ ] Style mappings (`@navikt/core/css/config/_mappings.js`)
- [ ] CSS class deprecations (`@navikt/aksel-stylelint/src/deprecations.ts`)
- [ ] Exports (`@navikt/core/react/src/index.ts` and `@navikt/core/react/package.json`)
- [ ] New component? CSS import (`@navikt/core/css/index.css`)
- [ ] Breaking change? Update migration guide. Consider codemod.
- [ ] Changeset (Format: `<Component>: <gitmoji?> <Text>` E.g. "Button: :sparkles: Add feature xyz")

### CODE REVIEW

File: .github/skills/ds-component-restructure/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+For each row in the mapping table where the path changes:
+
+1. Create the file at the new path with all import paths updated to reflect new locations
+2. Update every file in the "Importers to update" column — use the mapping table, don't rely on memory
+3. Do NOT delete old files yet
+
+### 4. Update `index.ts`
+
+The repo uses a mix of default and named exports because `react-docgen-typescript` (used by `scripts/docgen.ts`) has inconsistent pickup depending on export style. Follow the existing pattern in the component's original `index.ts` — do NOT change export style during restructure.
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md:88]: Not that it matters much for this skill, but are we sure this is why? Tried removing default export on Button, but docs were still generated 🤔

File: .github/skills/ds-component-restructure/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+Some components export everything as named (no `default`). If the original had no default export, keep it that way. If types are in `<Component>.types.ts`, re-export them from there directly or via the root.
+
+### 5. Remove Old Files
+
+Only after verifying new files are correct and `index.ts` exports match the original, delete the old files.
+
+### 6. Validate No Breaking Changes
+
+**Step 1 — Export diff.** Before deleting old files, extract all exported names from the original `index.ts` (recorded in step 1) and compare against the new `index.ts`. Every name must be present:
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: Step 5: "delete old files", step 6: "before deleting old files"... 🤔

File: .github/skills/ds-component-restructure/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+```bash
+# Exports in new index.ts that are missing from original → regression
+# Exports in original that are missing from new index.ts → missing export (breaking change)
+```
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: bash?

File: .github/skills/ds-component-restructure/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+**Step 2 — Build.** Run `yarn workspace @navikt/ds-react build` — must succeed with no TypeScript errors.
+
+**Step 3 — Package root.** Verify `@navikt/core/react/src/index.ts` still re-exports the component. No changes needed there unless the component's `index.ts` path changed (it shouldn't).
+
+### 7. Update Stories
+
+Move stories to component root dir. If stories import from internal paths, update those paths.
+
+**Exception:** A `stories/` subdirectory is acceptable if it already exists — do not flatten it.
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md:129]: Isn't this more or less covered by "Canonical Structure"/"Naming Rules"?

File: .github/skills/ds-component-restructure/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+Unit tests live next to the file they test — same directory, same base name with `.test.ts` / `.test.tsx` suffix.
+
+| What to test           | File location                                       |
+| ---------------------- | --------------------------------------------------- |
+| Helper utility         | `helpers/<helper-name>.test.ts`                     |
+| Sub-component logic    | `<subcomponent>/<Component><Subcomponent>.test.tsx` |
+| Root component / hooks | `root/<Component>Root.test.tsx`                     |
+| Context helpers        | `root/<Component>Root.context.test.ts`              |
+
+Do NOT move existing test files unless their source file moves. When moving a source file, move its test file with it and update import paths.
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md:142]: Isn't this more or less covered by "Canonical Structure"/"Naming Rules"?

File: .github/skills/ds-component-restructure/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+### Flat components
+
+Leave flat components flat — do NOT introduce subdirs unless the component contains internal utility-components (files that render JSX but are not public sub-components). If such a utility-component exists, move it to its own subdir following the canonical naming.
+
+Example: `button/Button.tsx` has no other sub-components, so should stay in root.
+
+### `parts/` directory
+
+If the component uses a `parts/` dir instead of `root/`, rename it to `root/` during restructure. Update all relative imports.
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: This doesn't make any sense to me. "parts" often contains sub-components, so it will lead to wrong dir structure. Do we need to mention it at all?

File: .github/skills/ds-component-restructure/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+| Root component / hooks | `root/<Component>Root.test.tsx`                     |
+| Context helpers        | `root/<Component>Root.context.test.ts`              |
+
+Do NOT move existing test files unless their source file moves. When moving a source file, move its test file with it and update import paths.
+
+## Edge Cases
+
+### Flat components
+
+Leave flat components flat — do NOT introduce subdirs unless the component contains internal utility-components (files that render JSX but are not public sub-components). If such a utility-component exists, move it to its own subdir following the canonical naming.
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: Unclear what "flat components" are (to me at least 😅)

File: .github/skills/ds-component-restructure/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+├── <Component>Root.context.ts
+└── use<Hook>.ts # public hook
+```
+
+Export from `index.ts` the same way as components:
+
+```ts
+export { use<Hook> } from "./root/use<Hook>";
+export type { Use<Hook>Return } from "./root/use<Hook>";
+```
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: Suggestion: Remove this and add an entry about hooks under "Canonical Structure"/"Naming Rules". The last part about exports should already be covered by the "no breaking changes" instruction.

File: .github/skills/ds-component-restructure/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+### Nested subcomponents
+
+Do NOT nest subcomponent dirs inside other subcomponent dirs. All sub-component dirs must be direct children of the component root:
+
+```
+# Wrong
+checkbox/
+└── input/
+    └── CheckboxInput.tsx
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]: The code does not actually show an example of nested subcomponent dirs.

File: .github/skills/ds-component-restructure/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+```
+
+### Naming Rules
+
+| What                   | Pattern                                 | Example                       |
+| ---------------------- | --------------------------------------- | ----------------------------- |
+| Root component file    | `<Component>Root.tsx`                   | `AccordionRoot.tsx`           |
+| Sub-component file     | `<Component><Subcomponent>.tsx`         | `AccordionItem.tsx`           |
+| Internal sub-component | `<Component><Subcomponent>Internal.tsx` | `DialogBackdropInternal.tsx`  |
+| Sub-component dir      | lowercase sub-component name            | `item/`                       |
```
Comment (HalvorHaugan) [inline .github/skills/ds-component-restructure/SKILL.md]:
```suggestion
| Sub-component dir      | kebab-case sub-component name            | `item/`                       |
```

File: .github/skills/ds-component-restructure/SKILL.md (hunk 1 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md:88]: Depends on if the component has sub-components etc also, so easier to just keep it as-is 🤞

File: .github/skills/ds-component-restructure/SKILL.md (hunk 2 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md]: Fixed

File: .github/skills/ds-component-restructure/SKILL.md (hunk 3 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md]: Fixed

File: .github/skills/ds-component-restructure/SKILL.md (hunk 4 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md:129]: Yes, but LLMs love to follow step-by-step guides with explicit guidance 🦮

File: .github/skills/ds-component-restructure/SKILL.md (hunk 5 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md:142]: Yes, but gives more explicit context for what we mean by "unit tests" so the LLM can avoid guessing 🤞

File: .github/skills/ds-component-restructure/SKILL.md (hunk 6 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md]: Changed to: 
> If the component uses a `parts/` dir, break it up into separate sub-component dirs.

Was based on testing that it seems to not understand how /parts worked

File: .github/skills/ds-component-restructure/SKILL.md (hunk 7 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md]: The example below gives some context
> Example: `button/Button.tsx` has no other sub-components, so should stay in root.

But renamed the heading to `### Flat components without sub-components`

File: .github/skills/ds-component-restructure/SKILL.md (hunk 4 above)
Comment (HalvorHaugan) [reply .github/skills/ds-component-restructure/SKILL.md:129]: Ok, but it seems illogical to have this as step 7, since it should already have happened in step 3?

Comment (HalvorHaugan) [reply .github/skills/ds-component-restructure/SKILL.md:129]: I guess it's fine. Some suggestions you can consider:
- Merge into step 3
- Move it after step 3 or 4
- Make it h2 with phrasing more like the "Unit Tests" section

**Review by HalvorHaugan — APPROVED** [review]

File: .github/skills/ds-component-restructure/SKILL.md (hunk 4 above)
Comment (KenAJoh) [reply .github/skills/ds-component-restructure/SKILL.md:129]: Fixed on main
